File: Algorithm/Python/322.coin-change.py
#
# @lc app=leetcode id=322 lang=python
#
# [322] Coin Change
#

# @lc code=start
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [-1] * (amount+1)
        dp[0] = 0
        for c in coins:
            for i in range(c, amount+1):
                if dp[i-c] >= 0:
                    dp[i] = dp[i-c]+1 if dp[i] < 0 else min(dp[i], dp[i-c]+1)
        return dp[amount]
        
# @lc code=end

File: Algorithm/C++/322.coin-change.cpp
/*
 * @lc app=leetcode id=322 lang=cpp
 *
 * [322] Coin Change
 */

// @lc code=start
#include <vector>

class Solution {
public:
    int coinChange(std::vector<int>& coins, int amount) {
        std::vector<int> dp(amount+1, -1);
        dp[0] = 0;
        for (int c : coins) {
            for (int i = c; i <= amount; i++) {
                if (dp[i-c] >= 0) {
                    if (dp[i] < 0) {
                        dp[i] = dp[i-c]+1;
                    } else {
                        dp[i] = std::min(dp[i], dp[i-c]+1);
                    }
                }
            }
        }
        return dp[amount];
    }
};
// @lc code=end

